cluster/cluster.go

```diff
@@ -0,0 +1,472 @@
+	}
+	delete(c.sparesByID, id)
+	c.spares = nodeSliceWithoutID(c.spares, id)
+	c.usedNodes[id] = spareNode
+	return spareNode, nil
+}
+
+func (c *svcCluster) AddSpecifiedNode(newNode node.ServiceNode) error {
+	c.Lock()
+	defer c.Unlock()
+
+	if !c.isSpareNodeWithLock(newNode) {
+		return fmt.Errorf("provided node is not a known spare")
+	}
+
+	_, err := c.addNodeFromListWithLock([]services.PlacementInstance{newNode.(services.PlacementInstance)})
+	return err
+}
+
+func (c *svcCluster) isSpareNodeWithLock(n node.ServiceNode) bool {
+	_, ok := c.sparesByID[n.ID()]
+	return ok
+}
+
+func (c *svcCluster) addNodeFromListWithLock(candidates []services.PlacementInstance) (node.ServiceNode, error) {
+	if c.status != ClusterStatusRunning && c.status != ClusterStatusSetup {
+		return nil, errClusterUnableToAlterPlacement
+	}
+
+	var (
+		psvc         = c.placementSvc
+		newPlacement services.ServicePlacement
+		usedInstance services.PlacementInstance
+	)
+	if err := c.opts.PlacementServiceRetrier().Attempt(func() error {
+		var internalErr error
+		newPlacement, usedInstance, internalErr = psvc.AddInstance(candidates)
+		return internalErr
+	}); err != nil {
+		return nil, err
+	}
+
+	setupNode, err := c.markSpareUsedWithLock(usedInstance)
+	if err != nil {
+		return nil, err
+	}
+
+	return setupNode, c.setPlacementWithLock(newPlacement)
+}
+
+func (c *svcCluster) AddNode() (node.ServiceNode, error) {
+	c.Lock()
+	defer c.Unlock()
+
+	numSpares := len(c.spares)
+	if numSpares < 1 {
+		return nil, errInsufficientCapacity
+	}
+
+	return c.addNodeFromListWithLock(c.sparesAsPlacementInstaceWithLock())
+}
+
+func (c *svcCluster) setPlacementWithLock(p services.ServicePlacement) error {
+	for _, instance := range p.Instances() {
+		// nb(prateek): update usedNodes with the new shards.
+		instanceID := instance.ID()
+		usedNode, ok := c.usedNodes[instanceID]
+		if ok {
+			usedNode.SetShards(instance.Shards())
+		}
+	}
+
+	c.placement = p
+	return nil
+}
+
+func (c *svcCluster) sparesAsPlacementInstaceWithLock() []services.PlacementInstance {
+	spares := make([]services.PlacementInstance, 0, len(c.spares))
+	for _, spare := range c.spares {
+		spares = append(spares, spare.(services.PlacementInstance))
+	}
+	return spares
+}
+
+func (c *svcCluster) RemoveNode(i node.ServiceNode) error {
+	c.Lock()
+	defer c.Unlock()
+
+	if c.status != ClusterStatusRunning && c.status != ClusterStatusSetup {
+		return errClusterUnableToAlterPlacement
+	}
+
+	usedNode, ok := c.usedNodes[i.ID()]
+	if !ok {
+		return errNodeNotInUse
+	}
+
+	var (
+		newPlacement services.ServicePlacement
+		psvc         = c.placementSvc
+	)
+	if err := c.opts.PlacementServiceRetrier().Attempt(func() error {
+		var internalErr error
+		newPlacement, internalErr = psvc.RemoveInstance(i.ID())
+		return internalErr
+	}); err != nil {
+		return err
+	}
+
+	// update removed instance from used -> spare
+	// nb(prateek): this omits modeling "leaving" shards on the node being removed
```

Comment:
But only for `m3em` right?  This won't affect the actual placement I believe no?